evm/src/bn254.rs

```diff
@@ -0,0 +1,743 @@
+}
+
+pub fn tangent(p: Curve, q: TwistedCurve) -> Fp12 {
+    let [px, py] = p;
+    let [qx, qy] = q;
+
+    let cx = neg_fp(mul_fp(U256::from(3), mul_fp(px, px)));
+    let cy = mul_fp(U256::from(2), py);
+
+    sparse_embed(
+        sub_fp(mul_fp(py, py), U256::from(9)),
+        mul_fp2(embed_fp2(cx), qx),
+        mul_fp2(embed_fp2(cy), qy),
+    )
+}
+
+pub fn cord(p1: Curve, p2: Curve, q: TwistedCurve) -> Fp12 {
+    let [p1x, p1y] = p1;
+    let [p2x, p2y] = p2;
+    let [qx, qy] = q;
+
+    let cx = sub_fp(p2y, p1y);
+    let cy = sub_fp(p1x, p2x);
+
+    sparse_embed(
+        sub_fp(mul_fp(p1y, p2x), mul_fp(p2y, p1x)),
+        mul_fp2(embed_fp2(cx), qx),
+        mul_fp2(embed_fp2(cy), qy),
+    )
+}
+
+fn tangent_slope(p: Curve) -> Fp {
+    let [px, py] = p;
+    let num = mul_fp(mul_fp(px, px), U256::from(3));
+    let denom = mul_fp(py, U256::from(2));
+    div_fp(num, denom)
+}
+
+fn cord_slope(p: Curve, q: Curve) -> Fp {
+    let [px, py] = p;
+    let [qx, qy] = q;
+    let num = sub_fp(qy, py);
+    let denom = sub_fp(qx, px);
+    div_fp(num, denom)
+}
+
+fn third_point(m: Fp, p: Curve, q: Curve) -> Curve {
+    let [px, py] = p;
+    let [qx, _] = q;
+    let ox = sub_fp(mul_fp(m, m), add_fp(px, qx));
+    let oy = sub_fp(mul_fp(m, sub_fp(px, ox)), py);
+    [ox, oy]
+}
+
+fn curve_add(p: Curve, q: Curve) -> Curve {
+    if p == q {
+        curve_double(p)
+    } else {
+        third_point(cord_slope(p, q), p, q)
+    }
+}
+
+fn curve_double(p: Curve) -> Curve {
+    third_point(tangent_slope(p), p, p)
+}
+
+pub fn miller_loop(p: Curve, q: TwistedCurve) -> Fp12 {
+    const EXP: [usize; 253] = [
+        1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1,
+        1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0,
+        1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0,
+        1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0,
+        1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0,
+        1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0,
+        0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0,
+        1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    ];
+
+    let mut o = p;
+    let mut acc = embed_fp12(U256::one());
+    let mut line;
+
+    for i in EXP {
+        acc = mul_fp12(acc, acc);
+        line = tangent(o, q);
+        acc = mul_fp12(line, acc);
+        o = curve_double(o);
+        if i != 0 {
+            line = cord(p, o, q);
+            acc = mul_fp12(line, acc);
+            o = curve_add(p, o);
+        }
+    }
+    acc
+}
+
+pub fn tate(p: Curve, q: TwistedCurve) -> Fp12 {
+    let mut out = miller_loop(p, q);
```

Comment:
This doesn't need to be `mut`. I'd use 5 separate variables, and take that opportunity to give them descriptive names.